packages/teams-js/src/private/files.ts

```diff
@@ -529,4 +593,206 @@ export namespace files {
+    }
+
+    sendMessageToParent('files.remove3PCloudStorageProvider', [logoutRequest], callback);
+  }
+
+  /**
+   * @hidden
+   * Hide from docs
+   *
+   * Initiates the 3P cloud storage provider file action (New / Rename / Upload / Download / Delete)
+   *
+   * @param cloudStorageProviderFileActionRequest 3P cloud storage provider request info
+   * @param callback Callback that will be triggered post 3P cloud storage action
+   */
+  export function perform3PCloudStorageProviderFileAction(
+    cloudStorageProviderFileActionRequest: I3PCloudStorageProviderRequest<
+      I3PCloudStorageProviderActionRequestContentType
+    >,
+    callback: (error?: SdkError) => void,
+  ): void {
+    ensureInitialized(FrameContexts.content);
+
+    if (!validateNonEmptyContent(cloudStorageProviderFileActionRequest, callback)) {
+      return;
+    }
+    if (!validateArgumentConsistency(cloudStorageProviderFileActionRequest, callback)) {
+      return;
+    }
+    if (!validateContentForSupportedActions(cloudStorageProviderFileActionRequest, callback)) {
+      return;
+    }
+
+    if (!callback) {
+      throw new Error('[files.perform3PCloudStorageProviderFileAction] callback cannot be null');
+    }
+
+    sendMessageToParent(
+      'files.perform3PCloudStorageProviderFileAction',
+      [cloudStorageProviderFileActionRequest],
+      callback,
+    );
+  }
+
+  function validateNonEmptyContent(
+    cloudStorageProviderFileActionRequest: I3PCloudStorageProviderRequest<
+      I3PCloudStorageProviderActionRequestContentType
+    >,
+    callback?: (err?: SdkError) => void,
+  ): boolean {
+    if (!(cloudStorageProviderFileActionRequest && cloudStorageProviderFileActionRequest.content)) {
+      if (callback) {
+        callback({
+          errorCode: ErrorCode.INVALID_ARGUMENTS,
+          message: '3P cloud storage provider request content is missing',
+        });
+      }
+      return false;
+    }
+    return true;
+  }
+
+  function validateArgumentConsistency(
+    cloudStorageProviderFileActionRequest: I3PCloudStorageProviderRequest<
+      I3PCloudStorageProviderActionRequestContentType
+    >,
+    callback?: (err?: SdkError) => void,
+  ): boolean {
+    if (!cloudStorageProviderFileActionRequest.content.action) {
+      if (callback) {
+        callback({
+          errorCode: ErrorCode.INVALID_ARGUMENTS,
+          message: '3P cloud storage provider action cannot be undefined',
+        });
+      }
+      return false;
+    }
+
+    if (!cloudStorageProviderFileActionRequest.content.providerCode) {
+      if (callback) {
+        callback({
+          errorCode: ErrorCode.INVALID_ARGUMENTS,
+          message: '3P cloud storage provider code cannot be undefined',
+        });
+      }
+      return false;
+    }
+
+    return true;
+  }
+
+  function validateContentForSupportedActions(
+    cloudStorageProviderFileActionRequest: I3PCloudStorageProviderRequest<
```

Comment:
I mentioned this above, but being more restrictive about parameters developers are allowed to pass in would make it so you don't have to do complicated verification of parameters here.

Reply:
After splitting action method for different scenarios, this point is addressed. Removed unnecessary verification code.